### PR TITLE
Adds more bitrunning antagonists + fixes (READY)

### LINKED ISSUE #79465: Bitrunner Console Randomise button doesn't actually choose a random sim
Reporting client version: 514.1589

## Round ID:
[217948](https://scrubby.melonmesa.com/round/217948)

## Testmerges:
- [Changes how halloween lighting works slightly](https://www.github.com/tgstation/tgstation/pull/79340)
- [Makes lighting objects objects again](https://www.github.com/tgstation/tgstation/pull/78857)
- [Extends Halloween](https://www.github.com/tgstation/tgstation/pull/79316)

## Reproduction:
I hit random five sims in a row and got Ashen Inferno all five times - This is a 1 in 1889568 chance if you have all maps available. Prior rounds have always chosen AI if you have at least 2 points for a map, Clown Planet if you only have one point, and Gondola Asteroid if you have no points.

### PR DESCRIPTION
**Pull request by jlsnow301**
## About The Pull Request
Reopened #78997
Larger patch for bitrunning that addresses a few issues.

- Two new antagonists: cyber tac and netguardian
- Quantum server emag opportunity
- Modular mob packs: Like random spawners, but for groups
- Antag spawning fixed: vdom antags now have up to a 10% chance to spawn based on domains loaded
- Virtual domains are no longer all fullbright by default, only the outdoorsy ones
- Actually deletes legion map file, since it was removed in #79424

<details>
<summary>images</summary>

The netguardian prime
dreamseeker_eKi7Mhv45s

robot_48_animation

The glitch effect - this mob is being mutated
dreamseeker_NE4j4rCoez

Cyber tac (t2 antagonist)
Screenshot 2023-10-27 211732

</details>

## Why It's Good For The Game
- Bitrunning antagonists are so incredibly rare that it's underwhelming to play as one for the solid second they offer if you even get the role
- Bitrunners had basically no traitor route to follow, they became assistants with black outfits

Fixes #79465

<details>
<summary>More info</summary>

Bitrunners don't have any type of traitor options. If they're made into traitors, there's nothing bitrunner related they can do, and their access is particularly bad so it's like they're a worse assistant. I've coupled this with the bitrunning antagonist system, which is now fixed.\. Bitrunners can now attempt to coax these entities to come onto the station, however they are not given any form of allegiance for doing so (and are quite counterable).

Previously, vdom antagonists relied on so many factors to spawn that it basically wouldn't happen. Now, it runs on the server each time there is a map loaded, with increasing probability as the round progresses. This builds up the list of spawnable antagonists, of which two are new, including an entirely new giant mech megafauna. This is the first "megafauna-esque" basic mob in the game. Its AI is bad, it's really only meant to be player controlled, but this does mean an admin can spawn them. Being mech, they are very counterable with ion rifles and the like.

Several refactors, rewrites, and overall bug fixes are included in this PR. 

Lastly, I added a framework for making bitrunner maps more random, the modular mob spawning system, which works in conjunction with random crate locations.
</details>

## Changelog
jlsnow301, infraredbaron
:cl:
add: Bitrunning Patch 1 features a host of changes!
add: Added randomized mobs to virtual domains, which will be indicated with a unique icon.
add: New emag interaction with the quantum server. Antags will spawn more frequently, and they can hack themselves onto the station. You have been warned.
add: Both living and dead players can now see which mob is going to spawn an antagonist in the vdom.
add: Two new vdom antagonists: Cyber Tac and the NetGuardian. These unlock at specific thresholds.
balance: You can no longer stack copies of the same ability with bitrunning disks.
balance: Some of the disk items have been replaced with stronger versions.
fix: You can no longer spy on crew using the advanced camera console on syndicate assault.
fix: Fixed the spawning mechanism of virtual domain antagonists. You should now have a chance of playing as one. This chance increases as more domains are completed.
fix: Vdom antagonists shouldn't spawn at the end of the run any longer.
fix: The preference for vdom antagonists has been changed to factor in the new types. Check your preferences!
fix: The quantum server will now show its balloon alerts to all observers.
fix: Random domains should be fully random again.
/:cl:

### CODE REVIEW

**Comment by Dawnseer** [comment]
You hear that Mr. Bitrunner? That is the sound of inevitability.

**Review by Wallemations — APPROVED** [review]
goddamn those net guardian sprites are clean

**Comment by DaydreamIQ** [comment]
Weird issue i've noticed is that the antag preferences panel for glitch always displays it as turned off, toggling it then closing the panel immediately resets it to the off state
image

**Comment by jlsnow301** [comment]
should be fixed with the last commit, works on local

**Comment by NoxVS** [comment]
The new antags (Or at least the tactical guy) still have sleeping carp, which locks them out of using the guns they spawn with.

**Comment by jlsnow301** [comment]
> The new antags (Or at least the tactical guy) still have sleeping carp, which locks them out of using the guns they spawn with.

can you open an issue for this? not really sure how that'd be possible, can't repro this on local

File: code/modules/bitrunning/antagonists/_parent.dm [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+		"Commence clean-up protocol on corrupt data.",
+		"Begin scan for aberrant code for termination.",
+		"Initiate lockdown on all rogue scripts.",
+		"Run integrity check and purge for digital disorder."
+	)
+	explanation_text = pick(explanation_texts)
+	return ..()
+
+/datum/objective/bitrunning_glitch_fluff/check_completion()
+	var/list/alive = GLOB.alive_player_list.Copy()
```
Comment (LemonInTheDark) [inline code/modules/bitrunning/antagonists/_parent.dm:56]: why copy?

File: code/modules/bitrunning/antagonists/cyber_tac.dm [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+	slowdown_active = 0.5
+	ui_theme = "terminal"
+	inbuilt_modules = list(/obj/item/mod/module/armor_booster)
+	allowed_suit_storage = list(
+		/obj/item/ammo_box,
+		/obj/item/ammo_casing,
+		/obj/item/restraints/handcuffs,
+		/obj/item/assembly/flash,
+	)
+	skins = list(
```
Comment (LemonInTheDark) [inline code/modules/bitrunning/antagonists/cyber_tac.dm:80]: bro this syntax I will fillet fikou

File: code/modules/bitrunning/components/digital_aura.dm [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+	if(!length(dimensions))
+		return ELEMENT_INCOMPATIBLE
+
+	switch(dimensions["width"])
+		if(32)
+			base_icon = 'icons/effects/bitrunning.dmi'
+		if(48)
+			base_icon = 'icons/effects/bitrunning_48.dmi'
+
+		if(64)
```
Comment (LemonInTheDark) [inline code/modules/bitrunning/components/digital_aura.dm:25]:
```suggestion
		if(64)
```

File: code/modules/bitrunning/components/digital_aura.dm [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+			base_icon = 'icons/effects/bitrunning_64.dmi'
+
+	redshift = mutable_appearance(base_icon, "redshift")
+	redshift.blend_mode = BLEND_MULTIPLY
+
+	glitch_effect = mutable_appearance(base_icon, "glitch", MUTATIONS_LAYER, alpha = 150)
+
+	thing.add_overlay(list(glitch_effect, redshift))
+	thing.alpha = 210
+	thing.set_light(2, l_color = LIGHT_COLOR_BUBBLEGUM, l_on = TRUE)
```
Comment (LemonInTheDark) [inline code/modules/bitrunning/components/digital_aura.dm:35]: what is this element meant to be applied to? what happens if it's applied to something with a light already?

File: code/modules/bitrunning/components/digital_aura.dm [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+	thing.update_appearance()
+
+/datum/element/digital_aura/Detach(datum/source)
+	. = ..()
+	var/atom/thing = source
+
+	thing.cut_overlay(glitch_effect, redshift)
```
Comment (LemonInTheDark) [inline code/modules/bitrunning/components/digital_aura.dm:42]: this will fail. your overlays mutate based off the parent's icon size.
are you sure this should be an element?

File: code/modules/bitrunning/components/glitch.dm [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+/datum/component/glitch/Initialize(obj/machinery/quantum_server/server, obj/machinery/byteforge/forge)
+	if(!isliving(parent))
+		return COMPONENT_INCOMPATIBLE
+
+	RegisterSignal(forge, COMSIG_MACHINERY_POWER_RESTORED, PROC_REF(on_forge_power_restored))
+	RegisterSignals(forge, list(COMSIG_MACHINERY_BROKEN, COMSIG_MACHINERY_POWER_LOST), PROC_REF(on_forge_broken))
+	forge_ref = WEAKREF(forge)
+
+	var/mob/living/owner = parent
+	server.spawned_threat_refs.Remove(WEAKREF(owner)) // so the server doesn't dust us
```
Comment (LemonInTheDark) [inline code/modules/bitrunning/components/glitch.dm:14]: this seems like something that should be a proc on the server imo

File: code/modules/bitrunning/components/glitch.dm [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+	owner.maxHealth = clamp(current_max, 200, 500)
+	owner.fully_heal()
+
+	owner.AddElement(/datum/element/digital_aura)
+
+/datum/component/glitch/RegisterWithParent()
+	RegisterSignals(parent, list(COMSIG_LIVING_STATUS_UNCONSCIOUS, COMSIG_LIVING_DEATH), PROC_REF(on_death))
+
+/datum/component/glitch/UnregisterFromParent()
+	UnregisterSignal(parent, COMSIG_LIVING_DEATH, PROC_REF(on_death))
```
Comment (LemonInTheDark) [inline code/modules/bitrunning/components/glitch.dm:29]: why are these here? do you plan on changing parent?

File: code/modules/bitrunning/components/glitch.dm [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+/// Sakujo
+/datum/component/glitch/proc/dust_mob()
+	if(QDELETED(parent))
+		return
+
+	var/mob/living/owner = parent
+	owner.dust()
+	qdel(src)
```
Comment (LemonInTheDark) [inline code/modules/bitrunning/components/glitch.dm:38]: is this required won't dusting the mob do this on its own? or are you just being safe (do we want to qdel this if the dust fails?)

File: code/modules/bitrunning/components/glitch.dm [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+	SIGNAL_HANDLER
+
+	var/mob/living/player = parent
+	var/atom/movable/screen/alert/bitrunning/alert = player.throw_alert(
+		ALERT_BITRUNNER_GLITCH,
+		/atom/movable/screen/alert/bitrunning,
+		new_master = source,
+	)
+	alert.name = "Source Broken"
+	alert.desc = "Our byteforge has been broken."
```
Comment (LemonInTheDark) [inline code/modules/bitrunning/components/glitch.dm:66]: subtype?

File: code/modules/bitrunning/components/npc_friendly.dm [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+		ROLE_GLITCH,
+		ROLE_SYNDICATE,
+	)
+
+/datum/element/npc_friendly/Detach(datum/source)
+	. = ..()
+
+	var/mob/living/player = source
+
+	player.faction.Remove(
```
Comment (LemonInTheDark) [inline code/modules/bitrunning/components/npc_friendly.dm:30]: couldn't this accidentally remove faction affiliation? faction system kinda sucks tbh are you sure we want a footgun like this?

File: code/modules/bitrunning/components/service_style.dm [hunk 11]
```diff
@@ -0,0 +1,30 @@
+/// Gives the appearance of being an agent
+/datum/element/service_style
```
Comment (LemonInTheDark) [inline code/modules/bitrunning/components/service_style.dm:2]: this is kinda a crummy element imo, it doesn't really do anything event based it's basically just a fancy proc on human. pagin ninja he'd know better

File: code/modules/bitrunning/components/npc_friendly.dm [hunk 12]
```diff
@@ -0,0 +1,42 @@
+/// Makes a mob friendly with most NPC factions
+/datum/element/npc_friendly
```
Comment (LemonInTheDark) [inline code/modules/bitrunning/components/npc_friendly.dm:2]: similar concern about this one tho not as bold

File: code/modules/bitrunning/components/virtual_mob.dm [hunk 13]
```diff
@@ -0,0 +1,38 @@
+
+/// Makes a carbon mob resist the elements, have no blood etc.
+/datum/element/virtual_mob
```
Comment (LemonInTheDark) [inline code/modules/bitrunning/components/virtual_mob.dm:3]: and this one

File: code/modules/bitrunning/event.dm [hunk 14]
```diff
@@ -7,16 +7,12 @@
...
-	weight = 10
-	/// List of active servers to choose from
-	var/list/obj/machinery/quantum_server/active_servers = list()
-	/// List of possible antags to spawn
-	var/static/list/possible_antags = list(
-		ROLE_CYBER_POLICE,
-	)
+	weight = 100
+	/// List of servers on the station
+	var/list/active_servers = list()
```
Comment (LemonInTheDark) [inline code/modules/bitrunning/event.dm:15]: is this list managed?

File: code/modules/bitrunning/virtual_domain/modular_mob_segment.dm [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+		return
+
+	var/total_amount = exact ? rand(1, max) : length(mobs)
+
+	shuffle_inplace(mobs)
+
+	var/turf/nearby = RANGE_TURFS(2, origin)
+	for(var/turf/tile in nearby)
+		if(tile.is_blocked_turf())
+			nearby -= tile
```
Comment (LemonInTheDark) [inline code/modules/bitrunning/virtual_domain/modular_mob_segment.dm:30]: may I suggest
```suggestion

	var/list/turf/nearby = list()
	for(var/turf/tile as anything in  RANGE_TURFS(2, origin))
		if(!tile.is_blocked_turf())
			nearby += tile
```

List subtraction is expensive, as anything because we know the type, and typing the list properly

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

File: code/modules/bitrunning/components/service_style.dm (hunk 11 above)
Comment (jlsnow301) [reply code/modules/bitrunning/components/service_style.dm:2]: it's entirely just a fancy proc on human, as are the others. however with this solution it just lets admins use addelement if they want

File: code/modules/bitrunning/virtual_domain/modular_mob_segment.dm [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+	if(!prob(probability))
+		return
+
+	var/total_amount = exact ? rand(1, max) : length(mobs)
+
+	shuffle_inplace(mobs)
+
+
+	var/list/turf/nearby = list()
+	for(var/turf/tile as anything in  RANGE_TURFS(2, origin))
```
Comment (LemonInTheDark) [inline code/modules/bitrunning/virtual_domain/modular_mob_segment.dm:29]: fuck
```suggestion
	for(var/turf/tile as anything in RANGE_TURFS(2, origin))
```

File: code/modules/bitrunning/antagonists/cyber_tac.dm (hunk 2 above)
Comment (ninjanomnom) [reply code/modules/bitrunning/antagonists/cyber_tac.dm:80]: I probably would have made this a list of types

File: code/modules/bitrunning/components/digital_aura.dm [hunk 17]
```diff
@@ -0,0 +1,45 @@
+/// Creates a digital effect around the target
+/datum/element/digital_aura
```
Comment (ninjanomnom) [inline code/modules/bitrunning/components/digital_aura.dm:2]: This should be just a generic effect in animations.dm instead of an element. You don't do anything particularly complex worth isolating, or anything that requires the effect be a separate object.

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/modules/bitrunning/components/npc_friendly.dm (hunk 10 above)
Comment (ninjanomnom) [reply code/modules/bitrunning/components/npc_friendly.dm:30]: We should at least be storing what factions were on it before and setting it back to that, though that will still cause issues

File: code/modules/bitrunning/components/service_style.dm (hunk 11 above)
Comment (ninjanomnom) [reply code/modules/bitrunning/components/service_style.dm:2]: Admins can also just call the proc though

Comment (ninjanomnom) [inline code/modules/bitrunning/components/service_style.dm:2]: This should be in the elements folder instead of in a module

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/modules/bitrunning/antagonists/cyber_tac.dm (hunk 2 above)
Comment (jlsnow301) [reply code/modules/bitrunning/antagonists/cyber_tac.dm:80]: it's pretty wild!

File: code/modules/bitrunning/components/glitch.dm (hunk 8 above)
Comment (jlsnow301) [reply code/modules/bitrunning/components/glitch.dm:38]: It does I'm just being super paranoid

File: code/modules/bitrunning/antagonists/_parent.dm (hunk 1 above)
Comment (jlsnow301) [reply code/modules/bitrunning/antagonists/_parent.dm:56]: I was going to mutate this, notably, removing the self from the list. It seemed silly enough of an objective that I overlooked completing that

File: code/modules/bitrunning/components/glitch.dm (hunk 9 above)
Comment (jlsnow301) [reply code/modules/bitrunning/components/glitch.dm:66]: I've been on a kick of just reusing the same alert - they can return to subtypes? 

(Later I'd like to just make name/desc an argument on throw_alert)

File: code/modules/bitrunning/event.dm (hunk 14 above)
Comment (jlsnow301) [reply code/modules/bitrunning/event.dm:15]: Wdym managed in this context?

File: code/modules/bitrunning/components/service_style.dm (hunk 11 above)
Comment (jlsnow301) [reply code/modules/bitrunning/components/service_style.dm:2]: You're gonna hate it here if you look around

File: code/modules/bitrunning/antagonists/_parent.dm (hunk 1 above)
Comment (LemonInTheDark) [reply code/modules/bitrunning/antagonists/_parent.dm:56]: you can do `for(var/thing in list - entry)` without mutating list, I'd suggest that here

File: code/modules/bitrunning/components/glitch.dm (hunk 8 above)
Comment (LemonInTheDark) [reply code/modules/bitrunning/components/glitch.dm:38]: would we want it qdel'd if the dust didn't work out?

File: code/modules/bitrunning/components/glitch.dm (hunk 9 above)
Comment (LemonInTheDark) [reply code/modules/bitrunning/components/glitch.dm:66]: ah right, nvm then

File: code/modules/bitrunning/event.dm (hunk 14 above)
Comment (LemonInTheDark) [reply code/modules/bitrunning/event.dm:15]: is it a list of objects that we are holding references to

Comment (jlsnow301) [reply code/modules/bitrunning/event.dm:15]: it's cut as soon as the event fires. i can make them weakrefs too

File: code/modules/bitrunning/components/glitch.dm (hunk 6 above)
Comment (jlsnow301) [reply code/modules/bitrunning/components/glitch.dm:14]: this felt like an unnecessary getter/setter pattern

File: code/modules/bitrunning/components/digital_aura.dm (hunk 4 above)
Comment (jlsnow301) [reply code/modules/bitrunning/components/digital_aura.dm:34]: this is getting set on mobs

**Comment by K4rlox** [comment]
2 issues
1. you secretly nerfed/buffed the tech disks without saying anything
image


2. you removed hats from pod loadouts, i can get that due to some outfits having head armor
**but why not disable hats of those outfits instead of all outfits?**
in my case, the mime doesnt get a beret, its a issue for me since it doenst give a french accent and if your character is bald, you roll a bald mime which looks really bad

**Comment by jlsnow301** [comment]
> 1. you secretly nerfed/buffed the tech disks without saying anything

That's in the changelog actually 

> **but why not disable hats of those outfits instead of all outfits?**

Good point. I had code that removed the armor from hats but it just looked so profane I removed hats, but forgot about mimes. I'll readd hats

**Comment by Cobbes** [comment]
After this got merged, I suddenly can't use my hands as a bit avatar. Is this a known issue, or something to do with this patch?
